src/third_party/wiredtiger/src/meta/meta_ckpt.c

```diff
@@ -240,6 +244,217 @@ __ckpt_compare_order(const void *a, const void *b)
+ * __ckpt_parse_attributes --
+ *  Parse the attributes associated with this checkpoint.
+ */
+static int
+__ckpt_parse_attributes(
+	WT_SESSION_IMPL *session, WT_CKPT *ckpt, char *s, uint64_t len)
+{
+	char *str_p;
+	char str[WT_CKPT_METADATA_LEN_MAX];
+	char *save_ptr_attrs;
+	char *save_ptr_attr;
+	char *attr;
+	char *key;
+	char *val;
+	uint64_t i;
+	uint64_t j;
+	uint64_t attr_idx;
+	uint64_t attr_keys_len;
+
+	static char *attr_keys[] = {
+		"addr",
+		"order",
+		"time",
+		"size",
+		"write_gen",
+	};
+	int attr_keys_found[] = { 0, 0, 0, 0, 0 };
+
+	/*
+	 * strtok_r modifies the input string, so we must make
+	 * a copy of it.
+	 */
+	WT_ASSERT(session, len < WT_CKPT_METADATA_LEN_MAX);
+	strncpy(str, s, len);
+	str[len] = '\0';
+
+	/*
+	 * Split the attribute string by comma and process each attribute.
+	 */
+	str_p = str;
+	attr_keys_len = sizeof(attr_keys) / sizeof(attr_keys[0]);
+	save_ptr_attrs = NULL;
+
+	/* Note: str_p must be NULL after the first iteration. */
+	for (i = 0; i < attr_keys_len; str_p = NULL, i++) {
+		attr = strtok_r(str_p, ",", &save_ptr_attrs);
+		if (attr == NULL)
+			goto err;
+
+		/* An attribute's key and val are separated by '='. */
+		save_ptr_attr = NULL;
+		key = strtok_r(attr, "=", &save_ptr_attr);
+		val = strtok_r(NULL, "=", &save_ptr_attr);
+
+		/*
+		 * We expect the attributes appear in the order specified in
+		 * "attr_keys".
+		 * This avoids unnecessary string comparisons in the typical
+		 * case, speeding up overall parsing time by ~10%.
+		 */
+		for (j = 0; j < attr_keys_len; j++) {
+			/*
+			 * Start from the attribute we expect at the i-th token,
+			 * but still parse successfully if the attributes are
+			 * written out of order.
+			 */
+			attr_idx = (i + j) % attr_keys_len;
+
+			if (strcmp(attr_keys[attr_idx], key) == 0) {
+				attr_keys_found[attr_idx]++;
+
+				WT_RET(__ckpt_parse_attribute_value(
+					session, ckpt, attr_idx, val));
+
+				break;
+			}
+		}
+	}
+
+	/* Sanity check to see what we have parsed every attribute. */
+	for (i = 0; i < attr_keys_len; i++) {
+		if (attr_keys_found[i] != 1)
+			goto err;
+	}
+
+	return (0);
+
+err:
+	WT_RET_MSG(session, WT_ERROR, "corrupted checkpoint list");
+}
+
+/*
+ * __ckpt_load_fast_parser --
+ *
+ *	Parses checkpoint information for a checkpoint of the given name.
+ *  "s" contains the metadata attributes such as "addr", and the
+ *  pointer should be advanced.
+ *  This is an alterative to __ckpt_load() implemented with
+ *  a faster parser.
+ */
+int
+__ckpt_load_fast_parser(WT_SESSION_IMPL *session, WT_CKPT *ckpt,
+	const char *name, char **s)
+{
+	char *paren;
+
+	paren = NULL;
+
+	*s = strchr(name + sizeof(WT_CKPT_NAME_PREFIX), '=');
+	if (*s == NULL)
+		goto err;
+
+	WT_RET(__wt_strndup(session, name, (int)(*s - name), &ckpt->name));
+	/*
+	 * Find the checkpoint metadata config string inside
+ 	 * the parentheses.
+	 *
+	 * Advance past "=" to look for the closing parenthesis.
+	 */
+	(*s)++;
+	paren = strpbrk(*s + 1, ")(");
```

Comment:
Do we need to be safe and check for *s being NULL before moving further ?

Reply:
Do you mean after L428, or before it?

There is a check in L418 for *s being NULL. After L418 I don't think it's possible for *s to be NULL right?

Reply:
But I think you are moving *s further with (*s)++ so it is not the same as in L418 anymore.

Reply:
As long as *s + 1 is not NULL, it won't segfault and should find the closing parenthesis if it is there. I don't think *s can be NULL here since we are just incrementing *s by 2.